### ISSUE TITLE
pretrain redpajama.py FSDP failure

### ISSUE THREAD

**Issue by xsyourpal** [Lightning-AI/lit-llama#424]
Pretraining with redpajama sample data with FSDP fails.   I am using a system with 8 A100 80GB GPUs

I tried running the example as described:
"python pretrain/redpajama.py --devices 4  --train_data_dir data/lit-redpajama-sample" 

I am getting the following 
error:fabric.py",  line 815, in _validate_setup
 raise RuntimeError(
RuntimeError: The `Fabric` requires the model and optimizer(s) to be set up separately. Create and set up the model first through `model = self.setup_model(model)`. Then create the optimizer and set it up: `optimizer = self.setup_optimizer(optimizer)`.
 ........
line 116, in main  model, optimizer = fabric.setup(model, optimizer)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
I am looking for help on making the necessary changes since it looks like this error may be well understood but the scripts have not been updated.  I kind of understand the error, but don't know how to go about making the changes.  Appreciate any pointers to documentation or example I can look up to resolve this would also be most appreciated.

**Comment by carmocca** [comment]
What is your `lightning` version? This shouldn't happen using master.

You can upgrade it by doing:

`pip uninstall -y lightning; pip install -r requirements.txt`

**Comment by xsyourpal** [comment]
I followed the instruction you provided above.   I am using an conda environment.  Previously my Lightning version was 2.0.4
After I ran the pip command.   My lightning version is 2.1.0dev and my torch version is 2.1.0.dev20230706
When I repeat the pretrain command "python pretrain/redpajama.py --devices 4 --train_data_dir data/lit-redpajama-sample" in this new environment,  I now get the following error:

>>>>>>>>
 File "/redpajama/lit-llama/lit_llama/packed_dataset.py", line 188, in _load_n_chunks
    self._datasets = [iter(el) for el in datasets]
    raise StopIteration
    for iter_num, train_data in enumerate(train_dataloader):
                      ^^^^^^^^
    raise StopIteration
RuntimeError: generator raised StopIteration
  File "/redpajama/lit-llama/lit_llama/packed_dataset.py", line 58, in __iter__
StopIteration

The above exception was the direct cause of the following exception:

StopIteration
Traceback (most recent call last):
.....    
raise StopIteration
return _run_component(component, cfg_init)
return _run_component(component, cfg_init)
StopIteration
<<<<<<<<<
Please advise.  Thank You

**Comment by xsyourpal** [comment]
I followed the instruction you provided above.   I am using an conda environment.  Previously my Lightning version was 2.0.4
After I ran the pip command.   My lightning version is 2.1.0dev and my torch version is 2.0.1
I created a new conda environment.  The pip command ran cleanly.

When I repeat the pretrain command "python pretrain/redpajama.py --devices 4 --train_data_dir data/lit-redpajama-sample" in this new environment,  I now get the following error, complete error output is in the attached file.   

>>>
packed_dataset.py", line 188, in _load_n_chunks
    raise StopIteration
StopIteration
<<<<<<<
[lit-llama-error01.txt](https://github.com/Lightning-AI/lit-llama/files/12044145/lit-llama-error01.txt)
